src/network/routing/lokinet_router.cpp

```diff
@@ -0,0 +1,648 @@
+#include "session/network/routing/lokinet_router.hpp"
+
+#include <fmt/ranges.h>
+#include <fmt/std.h>
+#include <oxenc/base64.h>
+
+#include <llarp/contact/router_id.hpp>
+#include <lokinet.hpp>
+#include <oxen/log.hpp>
+#include <oxen/log/format.hpp>
+
+#include "session/network/network_opt.hpp"
+#include "session/onionreq/builder.hpp"
+#include "session/onionreq/response_parser.hpp"
+
+using namespace oxen;
+using namespace session;
+using namespace session::network;
+using namespace std::literals;
+using namespace oxen::log::literals;
+
+namespace session::network {
+
+namespace {
+    auto cat = oxen::log::Cat("network");
+
+    static constexpr std::string_view PROXIED_REQUESTS_KEY{"proxied_requests"};
+
+    std::string pending_request_key(const network_destination& dest) {
+        std::optional<std::string> key;
+
+        std::visit(
+                [&key](auto&& arg) {
+                    using T = std::decay_t<decltype(arg)>;
+
+                    if constexpr (std::is_same_v<T, oxen::quic::RemoteAddress>) {
+                        key = oxenc::to_hex(arg.view_remote_key());
+                    } else if constexpr (std::is_same_v<T, service_node>) {
```

Comment:
This is personal preference, but I'd be tempted to combine these since both have an identical call:

```suggestion
                    if constexpr (std::is_same_v<T, oxen::quic::RemoteAddress> || std::is_same_v<T, service_node>) {
```